### PR TITLE
[1/2] Allow homeservers to send registration emails | Sending the email

### PR DESCRIPTION
**Pull request by anoadragon453**
~~Fixes https://github.com/matrix-org/synapse/issues/5833~~ Moved out to ~~https://github.com/matrix-org/synapse/pull/5863~~
Part of fixing https://github.com/matrix-org/synapse/issues/5751

Decouples the activity of sending registration emails and binding them to an identity server.

This PR simply sends the registration email, but clicking it does not approve the user for registration. That will come in PR #2.

Some of this makes use of existing stuff for sending password reset emails from Synapse. Some work was done to make that stuff even more generic.

Upgrade notes:
* There is a new top-level config option, `account_threepid_delegate` which defines the address of an identity server that you would like to send registration/password reset emails on your behalf.

  The option `email.trust_identity_server_for_password_resets` has been replaced by this. If you set `email.trust_identity_server_for_password_resets` in your config to `true`, please remove it and configure `account_threepid_delegate` instead. The [sample config](https://github.com/matrix-org/synapse/blob/anoa/reg_email_sending_email/docs/sample_config.yaml) has information on how to configure it.

TODO:
  - [x] Change email configuration to use `account_threepid_delegate`

Note: This PR does not allow homeservers to send emails when simply adding an email to your account. That will come after this and will be blocked on a new MSC.

Part [2/2] will be successfully completing the registration step when `/submit_token` is hit with the correct details, and clearing out the `password_servlet` flag stuff, which is no longer needed. Will be a much smaller PR than this one.

~~Requires https://github.com/matrix-org/synapse/pull/5863~~ has been merged into the base branch.
~~Requires https://github.com/matrix-org/synapse/pull/5876~~ has been merged into the base branch.

### CODE REVIEW

File: synapse/rest/client/v2_alpha/_base.py [hunk 1]
```diff
@@ -26,7 +26,7 @@
 logger = logging.getLogger(__name__)
 
 
-def client_patterns(path_regex, releases=(0,), unstable=True, v1=False):
+def client_patterns(*path_regexes, releases=(0,), unstable=True, v1=False):
```
Comment (anoadragon453) [inline synapse/rest/client/v2_alpha/_base.py]: This originally was going to be added functionality to assist this PR, but the need was removed.

This seems like a nice thing to have though (be able to specify multiple regexes to `client_patterns`) so we may want to leave it in anyways?

File: synapse/config/emailconfig.py [hunk 2]
```diff
@@ -74,19 +74,20 @@ def read_config(self, config, **kwargs):
             "renew_at"
         )
 
-        email_trust_identity_server_for_password_resets = email_config.get(
-            "trust_identity_server_for_password_resets", False
```
Comment (richvdh) [inline synapse/config/emailconfig.py]: Could we give some help to people who set `trust_identity_server_for_password_resets` ? Generally we try very hard not to break existing configs, and breaking them silently is worse yet. Perhaps throw a ConfigError if `trust_identity_server_for_password_resets` is set?

File: synapse/config/emailconfig.py [hunk 3]
```diff
@@ -125,40 +126,63 @@ def read_config(self, config, **kwargs):
...
+            )
+            self.email_registration_template_failure_html = email_config.get(
+                "registration_template_failure_html", "registration_failure.html"
             )
-            # This template does not support any replaceable variables, so we will
-            # read it from the disk once during setup
-            email_password_reset_success_template = email_config.get(
-                "password_reset_success_template", "password_reset_success.html"
+
+            # These templates do not support any placeholder variables, so we
```
Comment (richvdh) [inline synapse/config/emailconfig.py]: I'm kinda wondering if a better solution is to remove the special-casing and stick them through the template engine anyway.

File: synapse/rest/client/v2_alpha/_base.py (hunk 1 above)
Comment (richvdh) [reply synapse/rest/client/v2_alpha/_base.py]: I'm not keen on registering the same servlet at multiple paths anyway, so I'd rather it went away rather than encouraging such naughtiness.

File: synapse/push/mailer.py [hunk 4]
```diff
@@ -123,6 +124,73 @@ def __init__(self, hs, app_name, template_html, template_text):
...
+            # Generate a session id
+            session_id = random_string(16)
+
+        # Generate a new validation token
+        token = random_string(32)
+
+        # Send the mail with the link containing the token, client_secret
+        # and session_id
+        try:
+            yield self.send_password_reset_mail(email, token, client_secret, session_id)
```
Comment (richvdh) [inline synapse/push/mailer.py]: I'm a bit confused about why this function claims to be used for "registration purposes" but we are sending a password reset email?

File: synapse/push/mailer.py [hunk 5]
```diff
@@ -131,14 +199,11 @@ def send_password_reset_mail(self, email_address, token, client_secret, sid):
             email_address (str): Email address we're sending the password
                 reset to
             token (str): Unique token generated by the server to verify
-                password reset email was received
+                the email was received
             client_secret (str): Unique token generated by the client to
                 group together multiple email sending attempts
             sid (str): The generated session ID
         """
-        if email.utils.parseaddr(email_address)[1] == "":
```
Comment (richvdh) [inline synapse/push/mailer.py]: isn't this useful? we could move it to `send_email`?

File: synapse/config/registration.py [hunk 6]
```diff
@@ -99,6 +99,7 @@ def read_config(self, config, **kwargs):
         self.trusted_third_party_id_servers = config.get(
             "trusted_third_party_id_servers", ["matrix.org", "vector.im"]
         )
+        self.account_threepid_delegate = config.get("account_threepid_delegate")
```
Comment (richvdh) [inline synapse/config/registration.py]: this never appears to be read as anything other than a boolean - is proper support for it on its way?

File: synapse/push/mailer.py [hunk 7]
```diff
@@ -153,6 +218,33 @@ def send_password_reset_mail(self, email_address, token, client_secret, sid):
...
+            self.hs.config.public_baseurl
+            + "_matrix/client/unstable/registration/email/submit_token"
+            "?token=%s&client_secret=%s&sid=%s" % (token, client_secret, sid)
+        )
+
+        template_vars = {"link": link}
+
+        yield self.send_email(
+            email_address,
+            "[%s] Password Reset Email" % self.hs.config.server_name,
```
Comment (richvdh) [inline synapse/push/mailer.py]: not a password reset?

File: synapse/push/mailer.py [hunk 8]
```diff
@@ -153,6 +218,33 @@ def send_password_reset_mail(self, email_address, token, client_secret, sid):
...
+                link to
+            token (str): Unique token generated by the server to verify
+                the email was received
+            client_secret (str): Unique token generated by the client to
+                group together multiple email sending attempts
+            sid (str): The generated session ID
+        """
+        link = (
+            self.hs.config.public_baseurl
+            + "_matrix/client/unstable/registration/email/submit_token"
```
Comment (richvdh) [inline synapse/push/mailer.py]: we shouldn't be using /unstable/ for this (nor should we for password_reset, for that matter; we could fix that at the same time). Why don't we just use /r0/ to match the other endpoints?

File: synapse/rest/client/v2_alpha/account.py [hunk 9]
```diff
@@ -224,38 +155,45 @@ def on_POST(self, request):
...
         self.auth = hs.get_auth()
         self.config = hs.config
         self.clock = hs.get_clock()
         self.datastore = hs.get_datastore()
 
     @defer.inlineCallbacks
-    def on_GET(self, request, medium):
+    def on_GET(self, request, purpose, medium):
+        if purpose == "password_reset":
+            purpose = "password resets"
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: I think overwriting the input param is a bit confusing, and it doesn't really work (you end up with strings like "User registration have been disabled" and "Email-based password resets is disabled". Better to make each message conditional as a whole.

(In fact, it might end up being clearer to have two separate servlets and factor out the common code to a base class or to utility functions, but I don't know if that would work out cleaner or not)

File: synapse/rest/client/v2_alpha/_base.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/_base.py]: It came in handy for the submitToken endpoint for which the same code could be used for both email and msisdn. We could have two servlets that run the same method, though that would mean putting the method in another class/not in a class.

File: synapse/config/emailconfig.py [hunk 10]
```diff
@@ -125,40 +126,63 @@ def read_config(self, config, **kwargs):
...
+            # These email templates have placeholders in them, and thus must be
+            # parsed using a templating engine during a request
             self.email_password_reset_template_html = email_config.get(
                 "password_reset_template_html", "password_reset.html"
             )
             self.email_password_reset_template_text = email_config.get(
                 "password_reset_template_text", "password_reset.txt"
             )
-            self.email_password_reset_failure_template = email_config.get(
-                "password_reset_failure_template", "password_reset_failure.html"
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/config/emailconfig.py]: https://github.com/matrix-org/synapse/pull/5863

File: synapse/config/emailconfig.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/config/emailconfig.py]: It would be more clear, but we'd have to pull the template from the disk every time we get a registration request. So a question of code quality versus performance.

File: synapse/push/mailer.py (hunk 8 above)
Comment (anoadragon453) [reply synapse/push/mailer.py]: This was @erikjohnston's suggestion, and I think the reason was that this isn't a specced endpoint. It's simply left up to the implementation to decide.

But if we don't have a policy on unspecced things being under `/r0/` and/or want to reserve `/unstable/` for experimental stuff then we probably should move it.

File: synapse/push/mailer.py (hunk 5 above)
Comment (anoadragon453) [reply synapse/push/mailer.py]: Ah, this isn't clear from the diff at all but it's already in `send_email`:

https://github.com/matrix-org/synapse/blob/ce5f1cb98c164b5150e21d7c138f19153b4c40dd/synapse/push/mailer.py#L231-L242

File: synapse/push/mailer.py (hunk 8 above)
Comment (anoadragon453) [reply synapse/push/mailer.py]: Actually, we need to add another endpoint for just adding an email to your account. So this function should be split up into multiple servlets.

File: synapse/rest/client/v2_alpha/_base.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/_base.py]: We need to add another endpoint for just adding an email to your account. We should probably split this function up into multiple servlets.

File: changelog.d/5835.feature [hunk 11]
```diff
@@ -0,0 +1 @@
+Ability to send registration emails from the homeserver.
```
Comment (richvdh) [inline changelog.d/5835.feature]:
```suggestion
Add the ability to send registration emails from the homeserver rather than delegating to an Identity Server.
```

File: synapse/config/emailconfig.py [hunk 12]
```diff
@@ -146,28 +143,45 @@ def read_config(self, config, **kwargs):
...
+            # These email templates have placeholders in them, and thus must be
+            # parsed using a templating engine during a request
             self.email_password_reset_template_html = email_config.get(
                 "password_reset_template_html", "password_reset.html"
             )
             self.email_password_reset_template_text = email_config.get(
                 "password_reset_template_text", "password_reset.txt"
             )
+            self.email_registration_template_html = email_config.get(
+                "registration_template_html", "registration.html"
```
Comment (richvdh) [inline synapse/config/emailconfig.py]: IIRC we have to add new templates to UPGRADE.rst (and point the changelog to it), so that people who use a custom template_dir don't get messed up by it.

File: synapse/config/emailconfig.py [hunk 13]
```diff
@@ -146,28 +143,45 @@ def read_config(self, config, **kwargs):
...
+            )
+            self.email_registration_template_text = email_config.get(
+                "registration_template_text", "registration.txt"
```
Comment (richvdh) [inline synapse/config/emailconfig.py]: this should probably be in the sample config?

File: synapse/config/emailconfig.py (hunk 3 above)
Comment (richvdh) [reply synapse/config/emailconfig.py]: given we have to do that anyway for the other templates, it feels like a hit worth taking. But also, maybe something to think about another time.

File: synapse/rest/client/v2_alpha/account.py [hunk 14]
```diff
@@ -119,10 +119,14 @@ def on_POST(self, request):
                 send_attempt,
                 next_link,
             )
-        elif self.config.email_threepid_behaviour == ThreepidBehaviour.LOCAL:
+        elif self.config.threepid_behaviour == ThreepidBehaviour.LOCAL:
             # Send password reset emails from Synapse
-            sid = yield self.send_password_reset(
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: does this leave us with an unused `send_password_reset` method?

File: synapse/push/mailer.py [hunk 15]
```diff
@@ -123,6 +124,84 @@ def __init__(self, hs, app_name, template_html, template_text):
 
         logger.info("Created Mailer for app_name %s" % app_name)
 
+    @defer.inlineCallbacks
+    def send_threepid_validation(
```
Comment (richvdh) [inline synapse/push/mailer.py]: this doesn't really feel like it belongs in `Mailer`. In my head, `Mailer` is only concerned about formatting and sending emails; the session management stuff belongs elsewhere. I think `IdentityHandler` might be a better place.

File: synapse/rest/client/v2_alpha/account.py [hunk 16]
```diff
@@ -363,9 +367,7 @@ def load_jinja2_template(self, template_dir, template_filename, template_vars):
     @defer.inlineCallbacks
     def on_POST(self, request, medium):
         if medium != "email":
-            raise SynapseError(
-                400, "This medium is currently not supported for password resets"
-            )
+            raise SynapseError(400, "This medium is currently not supported")
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: Is this change necessary? It doesn't look like an improvement

File: synapse/rest/client/v2_alpha/account.py [hunk 17]
```diff
@@ -499,13 +501,23 @@ class EmailThreepidRequestTokenRestServlet(RestServlet):
...
+        self.config = hs.config
         self.store = self.hs.get_datastore()
+        self.identity_handler = hs.get_handlers().identity_handler
 
     @defer.inlineCallbacks
     def on_POST(self, request):
+        if self.config.threepid_behaviour == "off":
+            if self.config.local_threepid_emails_disabled_due_to_config:
+                logger.warn(
+                    "User password resets have been disabled due to lack of email config"
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: This seems to be in conflict with the error at line 518... is this servlet used for password resets, adding identities, or both?

File: synapse/config/emailconfig.py [hunk 18]
```diff
@@ -177,9 +191,15 @@ def read_config(self, config, **kwargs):
             filepath = os.path.join(
                 self.email_template_dir, email_password_reset_template_success_html
             )
-            self.email_password_reset_template_success_html_content = self.read_file(
+            self.email_password_reset_template_success_html = self.read_file(
```
Comment (richvdh) [inline synapse/config/emailconfig.py]: I think the old name was better tbh... it reflects the fact that this is the content of the template file, not (as per the other `*_template_html` vars) the path to the template file.

File: synapse/rest/client/v2_alpha/account.py [hunk 19]
```diff
@@ -499,13 +501,23 @@ class EmailThreepidRequestTokenRestServlet(RestServlet):
...
         super(EmailThreepidRequestTokenRestServlet, self).__init__()
-        self.identity_handler = hs.get_handlers().identity_handler
+        self.hs = hs
+        self.config = hs.config
         self.store = self.hs.get_datastore()
+        self.identity_handler = hs.get_handlers().identity_handler
 
     @defer.inlineCallbacks
     def on_POST(self, request):
+        if self.config.threepid_behaviour == "off":
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: ThreepidBehaviour.OFF ?

File: synapse/rest/client/v2_alpha/account.py [hunk 20]
```diff
@@ -542,8 +554,8 @@ class MsisdnThreepidRequestTokenRestServlet(RestServlet):
     def __init__(self, hs):
         self.hs = hs
         super(MsisdnThreepidRequestTokenRestServlet, self).__init__()
-        self.identity_handler = hs.get_handlers().identity_handler
         self.store = self.hs.get_datastore()
+        self.identity_handler = hs.get_handlers().identity_handler
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: nit: I'd prefer it if we could avoid no-op changes like this. It makes `git blame` less useful, and increases the likelihood of conflicts.

File: synapse/rest/client/v2_alpha/register.py [hunk 21]
```diff
@@ -73,8 +75,31 @@ def __init__(self, hs):
         self.hs = hs
         self.identity_handler = hs.get_handlers().identity_handler
 
+        if self.hs.config.threepid_behaviour == "local":
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: `ThreepidBehaviour.LOCAL`

File: synapse/rest/client/v2_alpha/register.py [hunk 22]
```diff
@@ -73,8 +75,31 @@ def __init__(self, hs):
...
+            self.mailer = Mailer(
+                hs=self.hs,
+                app_name=self.hs.config.email_app_name,
+                template_html=templates[0],
+                template_text=templates[1],
+            )
+
     @defer.inlineCallbacks
     def on_POST(self, request):
+        if self.hs.config.threepid_behaviour == "off":
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: ThreepidBehaviour.OFF

File: synapse/rest/client/v2_alpha/register.py [hunk 23]
```diff
@@ -188,6 +213,81 @@ def on_POST(self, request):
...
+        self.clock = hs.get_clock()
+        self.store = hs.get_datastore()
+
+    @defer.inlineCallbacks
+    def on_GET(self, request, purpose, medium):
+        if medium != "email":
+            raise SynapseError(
+                400, "This medium is currently not supported for registration"
+            )
+        if self.config.threepid_behaviour == "off":
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: OFF

File: synapse/config/emailconfig.py [hunk 24]
```diff
@@ -99,15 +99,12 @@ def read_config(self, config, **kwargs):
                 )
 
         self.local_threepid_emails_disabled_due_to_config = False
```
Comment (richvdh) [inline synapse/config/emailconfig.py]: I suppose this should be renamed, ideally.

File: synapse/rest/client/v2_alpha/account.py (hunk 19 above)
Comment (richvdh) [inline synapse/rest/client/v2_alpha/account.py]: given that we have about 15 copies of this pattern now, how about a method `self.config.assert_threepid_emails_enabled()` which does these checks and raises a generic SynapseError?

File: synapse/rest/client/v2_alpha/register.py [hunk 25]
```diff
@@ -188,6 +213,81 @@ def on_POST(self, request):
...
+            if self.config.local_threepid_emails_disabled_due_to_config:
+                logger.warn(
+                    "User registration has been disabled due to lack of email config"
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: not all registration, surely?

File: synapse/rest/client/v2_alpha/register.py [hunk 26]
```diff
@@ -188,6 +213,81 @@ def on_POST(self, request):
...
+                )
+            raise SynapseError(
+                400, "Email-based registration is disabled on this server"
+            )
+
+        sid = parse_string(request, "sid")
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: I think you need a `required=True` on these

File: synapse/rest/client/v2_alpha/register.py [hunk 27]
```diff
@@ -188,6 +213,81 @@ def on_POST(self, request):
...
+            html = self.load_jinja2_template(
+                self.config.email_template_dir,
+                self.config.email_registration_template_failure_html,
+                template_vars={"failure_reason": e.msg},
+            )
+            request.setResponseCode(e.code)
+
+        request.write(html.encode("utf-8"))
+        finish_request(request)
+        return None
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: this is redundant

File: synapse/rest/client/v2_alpha/register.py [hunk 28]
```diff
@@ -188,6 +213,81 @@ def on_POST(self, request):
         )
 
 
+class RegistrationSubmitTokenServlet(RestServlet):
+    """Handles registration 3PID validation token submission"""
+
+    PATTERNS = client_patterns(
+        "/registration/(?P<medium>[^/]*)/submit_token/*$", releases=(), unstable=True
```
Comment (richvdh) [inline synapse/rest/client/v2_alpha/register.py]: no need to accept trailing slashes
```suggestion
        "/registration/(?P<medium>[^/]*)/submit_token$", releases=(), unstable=True
```

File: changelog.d/5835.feature (hunk 11 above)
Comment (anoadragon453) [reply changelog.d/5835.feature]: FYI I think identity server is usually lowercase, whereas Identity Service API is uppercase.

File: synapse/config/emailconfig.py (hunk 12 above)
Comment (anoadragon453) [reply synapse/config/emailconfig.py]: Oh, hrm. Did not do that for password resets: https://github.com/matrix-org/synapse/blob/master/UPGRADE.rst#email

Should I add it retroactively in this PR as well as those for v1.4.0 templates?

File: synapse/rest/client/v2_alpha/account.py (hunk 17 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: Actually, this whole section shouldn't even have these checks as this PR should only deal with registration.

File: synapse/rest/client/v2_alpha/account.py (hunk 20 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: Good point, this is silly.

File: synapse/rest/client/v2_alpha/register.py (hunk 25 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/register.py]: No registration for you.

<sub>have fixed</sub>

File: synapse/rest/client/v2_alpha/account.py (hunk 19 above)
Comment (anoadragon453) [reply synapse/rest/client/v2_alpha/account.py]: We could, though we can't log in `*config` so we'd need to do the logging for `local_threepid_handling_disabled_due_to_email_config` outside of that method. So we'd need to check if this method raised an exception, caught it, logged out that "This endpoint was disabled due to a lack of email config block" then bubble the exception.

File: synapse/config/emailconfig.py (hunk 12 above)
Comment (richvdh) [reply synapse/config/emailconfig.py]: yeahh, I think that would be best.

File: changelog.d/5835.feature (hunk 11 above)
Comment (richvdh) [reply changelog.d/5835.feature]: I think you're right!

File: synapse/rest/client/v2_alpha/account.py (hunk 17 above)
Comment (richvdh) [reply synapse/rest/client/v2_alpha/account.py]: even better :)

File: synapse/rest/client/v2_alpha/account.py (hunk 19 above)
Comment (richvdh) [reply synapse/rest/client/v2_alpha/account.py]: oh, well, we could put it somewhere other than in `*config` then. anyway doesn't matter for now.

File: UPGRADE.rst [hunk 29]
```diff
@@ -49,6 +49,42 @@ returned by the Client-Server API:
     # configured on port 443.
     curl -kv https://<host.name>/_matrix/client/versions 2>&1 | grep "Server:"
 
+Upgrading to v1.4.0
```
Comment (richvdh) [inline UPGRADE.rst]: this looks great, by the way, but we probably need some more words about the fact that registration with email addresses won't work any more unless you twiddle some options? OTOH that sort of additional stuff should probably go in  separate PR so we can get this one landed.

File: UPGRADE.rst [hunk 30]
```diff
@@ -49,6 +49,42 @@ returned by the Client-Server API:
...
+password reset and registration tokens itself. Currently Synapse only supports sending emails,
+and does not have support for phone-based password reset or account registration. If Synapse is
+configured to handle these on its own, the ``email`` block of the config must be configured. If
+not, then password resets and registration via third-party IDs will be disabled.
+
+Email templates
+---------------
+
+If you have configured a custom email template directory with the ``email.template_dir``
+option, be aware that templates ``registration.html`` and ``registration.txt`` have been
```
Comment (richvdh) [inline UPGRADE.rst]: also the success and failure templates?

File: UPGRADE.rst [hunk 31]
```diff
@@ -49,6 +49,42 @@ returned by the Client-Server API:
...
+added, and Synapse will expect them to exist inside the configured template directory.
+
+To view the default email templates, see `synapse/res/templates
```
Comment (richvdh) [inline UPGRADE.rst]: s/email//

File: UPGRADE.rst [hunk 32]
```diff
@@ -49,6 +49,42 @@ returned by the Client-Server API:
...
+``account_threepid_delegate`` is not set to a domain, then Synapse will attempt to send
+password reset and registration tokens itself. Currently Synapse only supports sending emails,
+and does not have support for phone-based password reset or account registration. If Synapse is
+configured to handle these on its own, the ``email`` block of the config must be configured. If
+not, then password resets and registration via third-party IDs will be disabled.
+
+Email templates
+---------------
+
+If you have configured a custom email template directory with the ``email.template_dir``
```
Comment (richvdh) [inline UPGRADE.rst]: s/email//.

Speaking of which, the config option is in a stupid place (or we should have a separate one for the html templates), but let's punt that for now.

File: synapse/config/emailconfig.py [hunk 33]
```diff
@@ -87,7 +87,10 @@ def read_config(self, config, **kwargs):
...
         # identity server in the process.
         self.using_identity_server_from_trusted_list = False
-        if config.get("trust_identity_server_for_password_resets", False) is True:
+        if (
+            not self.account_threepid_delegate
+            and config.get("trust_identity_server_for_password_resets", False) is True
+        ):
             # Use the first entry in self.trusted_third_party_id_servers instead
             if self.trusted_third_party_id_servers:
                 self.account_threepid_delegate = self.trusted_third_party_id_servers[0]
```
Comment (richvdh) [inline synapse/config/emailconfig.py]: ftr: it feels dodgy to me that `account_threepid_delegate` can be set either here or in `RegistrationConfig`. It probably needs factoring out at some point.

File: UPGRADE.rst (hunk 29 above)
Comment (anoadragon453) [reply UPGRADE.rst]: Have updated the text

File: synapse/config/emailconfig.py (hunk 33 above)
Comment (anoadragon453) [reply synapse/config/emailconfig.py]: The whole emailconfig needs factoring at some point. I'll try to keep this in mind for when that happens. (I'll add a todo)

File: synapse/config/emailconfig.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/config/emailconfig.py]: I'll punt it for later.
